--- a/etc/inittab	2014-12-10 09:19:20.729343788 -0300
+++ b/etc/inittab	2015-03-19 09:45:39.653810243 -0300
@@ -31,8 +31,7 @@
 tty2::respawn:/sbin/getty 38400 tty2
 
 # Put a getty on the serial port
-#ttyS0::respawn:/sbin/getty -L ttyS0 115200 vt100 # UNSUPPORT GENERIC_SERIAL
-ttyS0::respawn:/bin/sh				  # AMLOGIC_GENERAL_SERIAL
+ttyS0::respawn:/sbin/getty -L ttyS0 115200 vt100
 
 # Logging junk
 ::sysinit:/bin/mkdir -p /var/log
