--- storage/browser/blob/internal_blob_data.h.orig
+++ storage/browser/blob/internal_blob_data.h
@@ -71,10 +71,10 @@
     void RemoveBlobFromShareableItems(const std::string& blob_uuid);
 
     // The builder is invalid after calling this method.
-    scoped_ptr<::storage::InternalBlobData> Build();
+    scoped_ptr< ::storage::InternalBlobData> Build();
 
    private:
-    scoped_ptr<::storage::InternalBlobData> data_;
+    scoped_ptr< ::storage::InternalBlobData> data_;
 
     DISALLOW_COPY_AND_ASSIGN(Builder);
   };
